### PR TITLE
[21301] Remove `WriterProxyData` from public APIs

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description

This PR makes WriterProxyData an implementation detail of the library.
Changes include:

* Moving `PublicationBuiltinTopicData` from `fastdds::dds:builtin` to `fastdds::dds`
* Extending `PublicationBuiltinTopicData` with additional fields to mimick those of `WriterProxyData`
* Writer discovery callbacks changed to receive a `reason` and a `PublicationBuiltinTopicData`
* Method `fill_discovery_data_from_cdr_message` now fills a `PublicationBuiltinTopicData`
* Struct `WriterDiscoveryInfo` has been removed
* `WriterProxyData.hpp` has been moved to `src`

Changes on related repos:
* [x] Discovery server: https://github.com/eProsima/Discovery-Server/pull/95
* [x] Python Bindings: https://github.com/eProsima/Fast-DDS-python/pull/163
* [x] ShapesDemo: https://github.com/eProsima/ShapesDemo/pull/164

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- _N/A_ Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- _N/A_ Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- __NO__ Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- __NO__ Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 
    - Related documentation PR: eProsima/Fast-DDS-docs#866
- _N/A_ Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: include/fastdds/rtps/builtin/data/PublicationBuiltinTopicData.hpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+struct PublicationBuiltinTopicData
+{
+    /// Builtin topic Key
+    BuiltinTopicKey_t key;
+
+    /// Builtin participant topic Key
+    BuiltinTopicKey_t participant_key;
+
+    /// Topic name
+    std::string topic_name;
```
Comment (Mario-DL) [inline include/fastdds/rtps/builtin/data/PublicationBuiltinTopicData.hpp:44]: I guess that making these fields as `fastcdr::string_255` makes no difference

**Review by Mario-DL — CHANGES_REQUESTED** [review]
Nice job in this PR.
The commit stack is well organised and facilitates the reading.

Leaving some minor suggestions

File: include/fastdds/rtps/writer/WriterDiscoveryInfo.hpp [hunk 2]
```diff
@@ -19,53 +19,26 @@
...
- */
-struct WriterDiscoveryInfo
-{
-public:
-
-    //!Enum DISCOVERY_STATUS, four different status for discovered writers.
-    //!@ingroup RTPS_MODULE
-    // *INDENT-OFF* : Does not understand the #if correctly and ends up removing the ;
-    //                at the end of the enum, which does not build.
+//! Enum WRITER_DISCOVERY_STATUS, four different status for discovered writers.
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/rtps/writer/WriterDiscoveryInfo.hpp:27]: This is to generate the doxygen summary for the enumeration. I changed `//!` into `///` to be consistent with other parts of the code.

File: include/fastdds/rtps/builtin/data/PublicationBuiltinTopicData.hpp (hunk 1 above)
Comment (MiguelCompany) [reply include/fastdds/rtps/builtin/data/PublicationBuiltinTopicData.hpp:44]: I think it makes sense to change it.

**Review by Mario-DL — DISMISSED** [review]
LGTM with green CI

**Review by Mario-DL — DISMISSED** [review]
LGTM with green CI

**Review by Mario-DL — APPROVED** [review]
LGTM !

**Pull request merged**